File: python/infinicore/ops/spmv.py
from infinicore.lib import _infinicore
from infinicore.tensor import Tensor


def spmv(
    alpha: Tensor,
    ap: Tensor,
    x: Tensor,
    beta: Tensor,
    out: Tensor,
    *,
    uplo: int = 0,
):
    _infinicore.spmv_(
        alpha._underlying,
        ap._underlying,
        x._underlying,
        beta._underlying,
        out._underlying,
        uplo,
    )

    return out

File: src/infinicore/ops/spmv/spmv.cc
#include "infinicore/ops/blas_spmv.hpp"

#include "../../utils.hpp"

namespace infinicore::op {

INFINICORE_GRAPH_OP_DISPATCHERS_IMPL(BlasSpmv);

BlasSpmv::BlasSpmv(const Tensor &alpha, const Tensor &ap, const Tensor &x, const Tensor &beta, Tensor y, int uplo) {
    INFINICORE_ASSERT_TENSORS_SAME_DEVICE(alpha, ap, x, beta, y);
    INFINICORE_GRAPH_OP_DISPATCH(y->device().getType(), alpha, ap, x, beta, y, uplo);
}

void BlasSpmv::execute(const Tensor &alpha, const Tensor &ap, const Tensor &x, const Tensor &beta, Tensor y, int uplo) {
    INFINICORE_GRAPH_OP_RECORD_OR_RUN(BlasSpmv, alpha, ap, x, beta, y, uplo);
}

void blas_spmv_(const Tensor &alpha, const Tensor &ap, const Tensor &x, const Tensor &beta, Tensor y, int uplo) {
    BlasSpmv::execute(alpha, ap, x, beta, y, uplo);
}

} // namespace infinicore::op
